File: src/commands/edit.js
const chalk = require("chalk");
const inquirer = require("inquirer");
const WritersEditor = require("../editor");
const projectManager = require("../utils/project");

async function editCommand(target, options) {
  try {
    // Check if in a writers project
    if (!projectManager.isWritersProject()) {
      console.log(
        chalk.red(
          '❌ Not a Writers project. Run "writers init" to initialize.',
        ),
      );
      return;
    }

    let filePath = null;

    // If target is provided, find the file
    if (target) {
      const fileInfo = await findFile(target);
      if (!fileInfo) {
        console.log(chalk.red(`❌ File not found: ${target}`));
        console.log(
          chalk.yellow('💡 Use "writers list" to see available files'),
        );
        console.log(chalk.yellow('💡 Use "writers new" to create a new file'));
        return;
      }
      filePath = fileInfo.path;
    } else {
      // Show file selection menu
      filePath = await selectFileToEdit();
      if (!filePath) {
        console.log(chalk.gray("No file selected for editing."));
        return;
      }
    }

    // Show ASCII art banner before launching the editor
    console.log(
      chalk.bold.cyan(`
        __  _____  __   _   ______ _    __________       ________    ____
       /  |/  /\ \/ /  / | / / __ \ |  / / ____/ /      / ____/ /   /  _/
      / /|_/ /  \  /  /  |/ / / / / | / / __/ / /      / /   / /    / /
     / /  / /   / /  / /|  / /_/ /| |/ / /___/ /___   / /___/ /____/ /
    /_/  /_/   /_/  /_/ |_/\____/ |___/_____/_____/   \____/_____/___/

${chalk.magenta("        ✦✦✦  Welcome to your CLI Novel Adventure!  ✦✦✦")}
`),
    );

    // Small delay to let user read the message
    await new Promise((resolve) => setTimeout(resolve, 1000));

    // Launch the editor
    const editor = new WritersEditor();
    await editor.launch(filePath);
  } catch (error) {
    console.error(chalk.red("❌ Error launching editor:"), error.message);
    process.exit(1);
  }
}

/**
 * Find file by name or partial match
 */
async function findFile(target) {
  const categories = [
    "chapters",
    "scenes",
    "characters",
    "shortstories",
    "notes",
  ];

  // Try exact match first
  for (const category of categories) {
    const files = await projectManager.getFiles(category);

    // Try exact name match
    let file = files.find(
      (f) => f.name === target || f.name.toLowerCase() === target.toLowerCase(),
    );

    if (file) {
      return { ...file, category };
    }

    // Try partial match
    file = files.find(
      (f) =>
        f.name.toLowerCase().includes(target.toLowerCase()) ||
        projectManager.sanitizeFileName(f.name) === target,
    );

    if (file) {
      return { ...file, category };
    }
  }

  // Try with common prefixes
  const prefixes = ["chapter", "scene", "character"];
  for (const prefix of prefixes) {
    if (target.startsWith(prefix)) {
      const name = target.substring(prefix.length).replace(/^\d+/, "").trim();
      const category = prefix + "s";
      const files = await projectManager.getFiles(category);

      const file = files.find((f) =>
        f.name.toLowerCase().includes(name.toLowerCase()),
      );

      if (file) {
        return { ...file, category };
      }
    }
  }

  return null;
}

/**
 * Show file selection menu
 */
async function selectFileToEdit() {
  const allFiles = [];

  // Get files from all categories
  const categories = [
    { name: "chapters", icon: "📖", color: "blue" },
    { name: "scenes", icon: "🎬", color: "green" },
    { name: "characters", icon: "👤", color: "yellow" },
    { name: "shortstories", icon: "📚", color: "cyan" },
    { name: "notes", icon: "📝", color: "magenta" },
  ];

  for (const category of categories) {
    const files = await projectManager.getFiles(category.name);
    for (const file of files) {
      const colorFn = chalk[category.color] || chalk.white;
      allFiles.push({
        name: `${category.icon} ${colorFn(category.name.slice(0, -1))}: ${file.name}`,
        value: file.path,
        short: file.name,
      });
    }
  }

  if (allFiles.length === 0) {
    console.log(chalk.yellow('No files found. Create one with "writers new"'));
    return null;
  }

  // Add option to create new file
  allFiles.push(new inquirer.Separator());
  allFiles.push({
    name: chalk.green("+ Create new file"),
    value: "CREATE_NEW",
  });
  allFiles.push({
    name: chalk.gray("Cancel"),
    value: null,
  });

  const { selectedFile } = await inquirer.prompt([
    {
      type: "list",
      name: "selectedFile",
      message: "Select a file to edit:",
      choices: allFiles,
      pageSize: 15,
    },
  ]);

  if (selectedFile === "CREATE_NEW") {
    return await createNewFileDialog();
  }

  return selectedFile;
}

/**
 * Show dialog to create a new file
 */
async function createNewFileDialog() {
  const questions = [
    {
      type: "list",
      name: "type",
      message: "What type of file would you like to create?",
      choices: [
        { name: "📖 Chapter", value: "chapter" },
        { name: "🎬 Scene", value: "scene" },
        { name: "👤 Character", value: "character" },
        { name: "📚 Short Story", value: "shortstory" },
        { name: "📝 Note", value: "note" },
      ],
    },
    {
      type: "input",
      name: "name",
      message: "Enter the name:",
      validate: (input) => {
        if (input.trim().length === 0) {
          return "Name cannot be empty";
        }
        return true;
      },
    },
  ];

  const answers = await inquirer.prompt(questions);

  try {
    // Create the file using the existing new command logic
    const newCommand = require("./new");
    const pluralType =
      answers.type === "shortstory" ? "shortstories" : answers.type + "s";
    const result = await projectManager.createFile(
      pluralType,
      answers.name,
      "", // No template for now
    );

    console.log(
      chalk.green(`✅ Created ${answers.type}: ${chalk.bold(result.name)}`),
    );
    return result.path;
  } catch (error) {
    console.error(chalk.red("❌ Error creating file:"), error.message);
    return null;
  }
}

module.exports = editCommand;

File: rust-cli/src/commands/edit.rs
use anyhow::Result;
use colored::*;
use dialoguer::{theme::ColorfulTheme, Select};
use std::path::PathBuf;

use crate::editor::WritersEditor;
use crate::utils::ProjectManager;

pub async fn run(target: Option<String>) -> Result<()> {
    // Check if in a writers project
    if !ProjectManager::is_writers_project() {
        println!(
            "{}",
            "❌ Not a Writers project. Run \"writers init\" to initialize.".red()
        );
        return Ok(());
    }

    let target_file = if let Some(target) = target {
        // Try to find the specified file
        match ProjectManager::find_file(&target).await? {
            Some(file) => file,
            None => {
                println!("{} {}", "❌ File not found:".red(), target);
                println!(
                    "{}",
                    "💡 Use \"writers list\" to see available files".yellow()
                );
                return Ok(());
            }
        }
    } else {
        // Show file selection menu
        select_file_to_edit().await?
    };

    // Launch the built-in editor
    launch_editor(target_file).await
}

async fn select_file_to_edit() -> Result<PathBuf> {
    println!("{}", "📝 Select a file to edit:".cyan().bold());
    println!();

    let config = ProjectManager::get_config().await?;
    let mut all_files = Vec::new();
    let mut file_descriptions = Vec::new();

    // Get files based on project type
    match config.r#type.as_str() {
        "novel" => {
            // Add chapters
            let chapters = ProjectManager::get_chapters().await?;
            for chapter in chapters {
                if let Some(file_name) = chapter.file_name().and_then(|n| n.to_str()) {
                    all_files.push(chapter.clone());
                    file_descriptions.push(format!("📖 Chapter: {}", file_name));
                }
            }

            // Add scenes
            let scenes = ProjectManager::get_scenes().await?;
            for scene in scenes {
                if let Some(file_name) = scene.file_name().and_then(|n| n.to_str()) {
                    all_files.push(scene.clone());
                    file_descriptions.push(format!("🎬 Scene: {}", file_name));
                }
            }

            // Add characters
            let characters = ProjectManager::get_characters().await?;
            for character in characters {
                if let Some(file_name) = character.file_name().and_then(|n| n.to_str()) {
                    all_files.push(character.clone());
                    file_descriptions.push(format!("👤 Character: {}", file_name));
                }
            }

            // Add outline if it exists
            if std::path::Path::new("outline.md").exists() {
                all_files.push(PathBuf::from("outline.md"));
                file_descriptions.push("📋 Outline".to_string());
            }
        }
        "short-story" => {
            let stories = ProjectManager::get_short_stories().await?;
            for story in stories {
                if let Some(file_name) = story.file_name().and_then(|n| n.to_str()) {
                    all_files.push(story.clone());
                    file_descriptions.push(format!("📚 Story: {}", file_name));
                }
            }

            let characters = ProjectManager::get_characters().await?;
            for character in characters {
                if let Some(file_name) = character.file_name().and_then(|n| n.to_str()) {
                    all_files.push(character.clone());
                    file_descriptions.push(format!("👤 Character: {}", file_name));
                }
            }
        }
        "simple-short-story" => {
            let stories = ProjectManager::get_short_stories().await?;
            for story in stories {
                if let Some(file_name) = story.file_name().and_then(|n| n.to_str()) {
                    all_files.push(story.clone());
                    file_descriptions.push(format!("✍️  Draft: {}", file_name));
                }
            }
        }
        "blog" => {
            let posts = ProjectManager::get_blog_posts().await?;
            for post in posts {
                if let Some(file_name) = post.file_name().and_then(|n| n.to_str()) {
                    all_files.push(post.clone());
                    file_descriptions.push(format!("📝 Post: {}", file_name));
                }
            }
        }
        _ => {
            // Generic project - get all markdown files
            let all_content = ProjectManager::get_all_content_files().await?;
            for file in all_content {
                if let Some(file_name) = file.file_name().and_then(|n| n.to_str()) {
                    all_files.push(file.clone());
                    file_descriptions.push(format!("📄 File: {}", file_name));
                }
            }
        }
    }

    // Add notes
    let notes = ProjectManager::get_notes().await?;
    for note in notes {
        if let Some(file_name) = note.file_name().and_then(|n| n.to_str()) {
            all_files.push(note.clone());
            file_descriptions.push(format!("📝 Note: {}", file_name));
        }
    }

    if all_files.is_empty() {
        println!("{}", "No files found to edit.".yellow());
        println!("{}", "💡 Create some content first".yellow());
        return Err(anyhow::anyhow!("No files available"));
    }

    // Add README if it exists
    if std::path::Path::new("README.md").exists() {
        all_files.push(PathBuf::from("README.md"));
        file_descriptions.push("📖 README".to_string());
    }

    let selection = Select::with_theme(&ColorfulTheme::default())
        .with_prompt("Choose a file to edit")
        .default(0)
        .items(&file_descriptions)
        .interact()?;

    Ok(all_files[selection].clone())
}

async fn launch_editor(file_path: PathBuf) -> Result<()> {
    println!();
    println!(
        "{} {}",
        "🚀 Launching editor for:".green(),
        file_path.display().to_string().cyan()
    );
    println!();
    println!("{}", "Editor Controls:".yellow().bold());
    println!(
        "  • {} - Switch between Navigation and Insert mode",
        "ESC".cyan()
    );
    println!("  • {} - Save file", "Ctrl+S".cyan());
    println!("  • {} - Quit editor", "Ctrl+Q".cyan());
    println!("  • {} - Search text", "/".cyan());
    println!("  • {} - Toggle typewriter mode", "Ctrl+T".cyan());
    println!("  • {} - Toggle distraction-free mode", "F3".cyan());
    println!("  • {} - Undo", "Ctrl+Z".cyan());
    println!("  • {} - Redo", "Ctrl+Y".cyan());
    println!();
    println!("{}", "Navigation Mode:".yellow().bold());
    println!(
        "  • {} - Move cursor",
        "hjkl, Shift+WASD, or arrow keys".cyan()
    );
    println!("  • {} - Enter insert mode", "i".cyan());
    println!("  • {} - Enter insert mode at end of line", "a".cyan());
    println!("  • {} - Create new line and enter insert mode", "o".cyan());
    println!("  • {} - Delete current line", "dd".cyan());
    println!("  • {} - Go to beginning/end of line", "0 / $".cyan());
    println!("  • {} - Page up/down", "Page Up/Down".cyan());
    println!();
    println!("{}", "Insert Mode:".yellow().bold());
    println!("  • {} - Type normally", "Any character".cyan());
    println!("  • {} - Return to navigation mode", "ESC".cyan());
    println!(
        "  • {} - Move cursor while in insert mode",
        "Arrow keys".cyan()
    );
    println!();
    println!("{}", "Press any key to continue...".bright_black());

    // Wait for a key press (simple implementation)
    use std::io::{self, Read};
    let mut buffer = [0];
    let _ = io::stdin().read(&mut buffer);

    // Clear screen and launch editor
    print!("\x1B[2J\x1B[1;1H"); // Clear screen and move cursor to top-left

    let mut editor = WritersEditor::new();
    let file_path_str = file_path.to_string_lossy().to_string();

    match editor.launch(Some(file_path_str)).await {
        Ok(_) => {
            println!();
            println!("{}", "✅ Editor session completed.".green());
        }
        Err(e) => {
            println!();
            if e.to_string().contains("TTY") {
                println!("{}", "❌ Editor requires a real terminal to run.".red());
                println!("{}", "💡 Make sure you're running this command directly in a terminal, not through pipes or scripts.".yellow());
            } else {
                println!("{} {}", "❌ Editor error:".red(), e);
            }
            return Err(e);
        }
    }

    Ok(())
}
